### DISCUSSION TITLE
One key triggers rerender of components using another key

### DISCUSSION THREAD

**Discussion by nihiluis** [47ng/nuqs#627] [Q&A] [answered]
Hi!

Say I'm having `const [recipeId, setRecipeId] = useQueryState("id")` in RecipeComponent.tsx
and `const [age, setAge] = useQueryState("age")` in AgeComponent.tsx

If I use setAge in my AgeComponent, then it seems that my  RecipeComponent gets rerendered.

Is it correct like that?

Thank you.

**Comment by franky47** [comment, marked as answer]
Unfortunately this is a Next.js issue: `useQueryStates` calls `useSearchParams` internally, and it will re-render when _any_ search param changes.

**Reply by AshotN** [reply]
Will v2 be addressing this issue?

**Reply by franky47** [reply]
No, a change in Next.js core would be necessary to impact this behaviour.

**Reply by AshotN** [reply]
Is there any recommended work around in the mean time?

**Reply by franky47** [reply]
- Keep re-renders small by moving the useQueryState(s) hooks as deep in the tree as possible
- Update to `nuqs@latest` where objects and arrays are now referentially stable if the value for a key didn't change (preventing unnecessary useEffect runs)

**Reply by AshotN** [reply]
Hey @franky47 one more question if you don't mind

Wouldn't using the browser's native history API be a solution, mentioned in the [Next docs](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#using-the-native-history-api) here

EDIT: Or would this still cause `useSearchParams` to re-render any component that uses it?

If that's the case, would it be possible to write a hook that only subscribes to certain changes, similar to how Zustand operates?

**Reply by franky47** [reply]
nuqs uses the history API pattern described by the Next.js docs (`shallow: true`) by default, and only opts-in to using the router API when needing to recompute server-side code (when the server is interested in getting the updated search params, via `shallow: false`).

Either way, the way the Next.js router cores are organised makes it impossible to get fine-grained reactivity on a particular search param (or set of). We'd have to not rely on `useSearchParams`, which might cause a lot of issues. I might give it a try at some point though, but it's not a high priority on the roadmap at the moment.

**Reply by AshotN** [reply]
Understood, I think I'll spend some time diving into how nuqs works and see if I can understand how it all works. Thank you

**Reply by AshotN** [reply]
Just playing around, I see how annoying it is dealing with query params with JavaScript. 

I made a [fork](https://github.com/AshotN/nuqs) as an attempt, and [deployed](https://nuqs-i08x64gz9-ashotns-projects.vercel.app/playground/basic-counter) it

I thought I was making progress with limiting renders until I realized a regression when you have a link that goes to the page you are already on.

The query params get removed from the URL and that is hard to detect without the re-rendering :(

**Reply by franky47** [reply]
Honestly I think the biggest issue with not using `useSearchParams` is server-side rendering. I don't know how else you might get the search params from the request in a manner that's correct and doesn't trigger hydration issues.

**Reply by AshotN** [reply]
Do you mean during the initial render of the page? During SSR?

Correct me if I'm wrong, but server components don't rerun SSR during query param change.

**Reply by franky47** [reply]
If you use the `searchParams` page prop or anything that marks the page as dynamic (this is changing a bit in Next 15), and `shallow: false` in your hook to use the router rather than the History API, they do (and so do the client components that use useSearchParams).

**Reply by AshotN** [reply]
Hey, was wondering. Has Next 15 changed anything to resolve this problem?

**Reply by franky47** [reply]
No, it still uses a single Context to propagate the search params in `useSearchParams` ([source](https://github.com/vercel/next.js/blob/c8ac0bdd609561e61a6839c5ef6124ee41b56889/packages/next/src/client/components/app-router.tsx#L645C8-L645C9)). Our best hope would be for React to land [useContextSelector](https://github.com/reactjs/rfcs/pull/119), but there hasn't been movement on this front for a while.
